### PR TITLE
Immutable ShardRouting

### PR DESCRIPTION
**Pull request by ywelsch**
This PR makes ShardRouting immutable, which lets us get rid of defensive copying of all ShardRouting objects whenever we do a reroute. With immutable objects we only make copies for shard routings that change.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java [hunk 1]
```diff
@@ -26,23 +26,22 @@
...
 
 import java.io.IOException;
 
 /**
  */
 public class ClusterSearchShardsGroup implements Streamable, ToXContent {
 
-    private Index index;
-    private int shardId;
+    private ShardId shardId;
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java]: Could this be made `final`, and have a constructor that takes a `StreamInput`, and `readFrom` just delegate to the constructor?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java]: Not for `Streamable`. If you switch it to `Writeable` then yes. But that might make sense to wait for another PR? I don't know, I haven't been following this one very closely.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 2]
```diff
@@ -424,39 +419,41 @@ public String prettyPrint() {
...
+        addRecovery(initializedShard);
+        assignedShardsAdd(initializedShard);
+        return initializedShard;
     }
 
     /**
      * Relocate a shard to another node, adding the target initializing
      * shard as well as assigning it. And returning the target initializing
      * shard.
      */
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: Javadocs here need to be updated to say returning both the source and target shard.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java]: I chose not to convert ClusterSearchShardsGroup to Writeable in this PR to keep things focused on ShardRouting / RoutingNodes.

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/shards/ClusterSearchShardsGroup.java]: Makes sense

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java [hunk 3]
```diff
@@ -37,25 +40,35 @@
...
 
+    // public for tests
+    public RoutingNode(String nodeId, DiscoveryNode node, Iterable<ShardRouting> shardsCollection) {
+        this(nodeId, node);
+        for (ShardRouting shardRouting : shardsCollection) {
+            ShardRouting previousValue = shards.put(shardRouting.shardId(), shardRouting);
+            assert previousValue == null : "Cannot have two different shards with same shard id on same node";
+        }
+
+    }
```
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: Not sure if I see the utility in keeping this constructor around.  I saw 6 instances where this constructor is used in the tests.  May be better to get rid of this constructor and just create a map instead for those 6 instances, making entries shardId->shardRouting?  Or just have a constructor that takes an `Iterable<ShardRouting>` and automatically creates the internal map out of it in the constructor?

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: ok, I've instead generalized the earlier constructor by adding a varargs ShardRouting argument.

I've also changed the signature of the constructor used by RoutingNodes to make explicit that we need a LinkedHashMap preserving iteration order (and made that one package visible).

**Comment by ywelsch** [comment]
I've added a small commit that makes UnassignedInfo immutable as well (06e774f).

**Comment by ywelsch** [comment]
@s1monw Can you have a look at this one?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java [hunk 4]
```diff
@@ -37,25 +40,34 @@
...
         this.nodeId = nodeId;
         this.node = node;
         this.shards = shards;
     }
 
+    private static LinkedHashMap<ShardId, ShardRouting> buildShardRoutingMap(ShardRouting... shardRoutings) {
+        final LinkedHashMap<ShardId, ShardRouting> shards = new LinkedHashMap<>();
+        for (ShardRouting shardRouting : shardRoutings) {
+            ShardRouting previousValue = shards.put(shardRouting.shardId(), shardRouting);
+            assert previousValue == null : "Cannot have two different shards with same shard id on same node";
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: can we make this a hard check?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 5]
```diff
@@ -586,15 +583,27 @@ public RoutingNodeIterator routingNodeIter(String nodeId) {
...
-        candidate.reinitializeShard();
+        ShardRouting reinitializedShard = candidate.reinitializeShard();
+        updateAssigned(candidate, reinitializedShard);
         inactivePrimaryCount++;
         inactiveShardCount++;
+        return reinitializedShard;
+    }
 
+    private void updateAssigned(ShardRouting oldShard, ShardRouting newShard) {
+        assert oldShard.shardId().equals(newShard.shardId());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: can we have some messages for these asserts, they are a pain if they trip without a message

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 6]
```diff
@@ -983,6 +1021,38 @@ private void ensureMutable() {
         }
     }
 
+    public Iterator<ShardRouting> nodeInterleavedShardIterator() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: can you add javadocs for this?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 7]
```diff
@@ -983,6 +1021,38 @@ private void ensureMutable() {
...
+        final Queue<Iterator<ShardRouting>> queue = new ArrayDeque<>();
+        for (Map.Entry<String, RoutingNode> entry : nodesToShards.entrySet()) {
+            queue.add(entry.getValue().copyShards().iterator());
+        }
+        return new Iterator<ShardRouting>() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: can we use `org.elasticsearch.common.collect.Iterators#concat` for this?

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 8]
```diff
@@ -38,43 +38,31 @@
...
-    private Index index;
-    private int shardId;
-    private String currentNodeId;
-    private String relocatingNodeId;
-    private boolean primary;
-    private ShardRoutingState state;
-    private RestoreSource restoreSource;
-    private UnassignedInfo unassignedInfo;
-    private AllocationId allocationId;
+    private final ShardId shardId;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: good!

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 9]
```diff
@@ -636,21 +545,18 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        if (frozen && usePreComputedHashCode) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: can we just pre-compute the hashCode in the ctor instead? I can already see this going bad with concurrency and we might be just lucky we didn't hit that yet? Either we do in the ctor and make it final or we need a volatile read here

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 10]
```diff
@@ -278,57 +255,38 @@ public boolean allocatedPostIndexCreate(IndexMetaData indexMetaData) {
...
+        return in.readWriteable(ShardRouting::new);
     }
 
-    public static ShardRouting readShardRoutingEntry(StreamInput in, Index index, int shardId) throws IOException {
-        ShardRouting entry = new ShardRouting();
-        entry.readFrom(in, index, shardId);
-        return entry;
+    public static ShardRouting readShardRoutingEntry(StreamInput in, ShardId shardId) throws IOException {
+        return in.readWriteable(streamInput -> new ShardRouting(shardId, streamInput));
     }
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: hmm can't this just be replaced by 
```Java
return new ShardRouting(shardId, in);
```

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 11]
```diff
@@ -278,57 +255,38 @@ public boolean allocatedPostIndexCreate(IndexMetaData indexMetaData) {
...
     public ShardIterator shardsIt() {
-        return new PlainShardIterator(shardId(), asList);
+        return new PlainShardIterator(shardId, asList);
     }
 
     public static ShardRouting readShardRoutingEntry(StreamInput in) throws IOException {
-        ShardRouting entry = new ShardRouting();
-        entry.readFrom(in);
-        return entry;
+        return in.readWriteable(ShardRouting::new);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: why not:
```Java
return new ShardRouting(in);
```

also maybe just make that ctor public?

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 12]
```diff
@@ -607,17 +607,21 @@ public DateTimeZone readOptionalTimeZone() throws IOException {
...
-                        + "] returned null which is not allowed and probably means it screwed up the stream.");
-            }
-            return t;
+            return readWriteable(reader);
         } else {
             return null;
         }
     }
 
+    public <T extends Writeable> T readWriteable(Writeable.Reader<T> reader) throws IOException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I think that is obsolete if we fix the usage of it?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 13]
```diff
@@ -768,7 +769,7 @@ protected void runReplicateTest(ClusterState state, IndexShardRoutingTable shard
                     assertEquals(1, shardFailedRequests.length);
                     CapturingTransport.CapturedRequest shardFailedRequest = shardFailedRequests[0];
                     // get the shard the request was sent to
-                    ShardRouting routing = clusterService.state().getRoutingNodes().node(capturedRequest.node.getId()).get(request.shardId.id());
+                    ShardRouting routing = clusterService.state().getRoutingNodes().node(capturedRequest.node.getId()).copyShards().get(request.shardId.id());
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: why do we need to copy here?

**Comment by s1monw** [comment]
@ywelsch looks great - left some minors

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java (hunk 7 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: `Iterators#concat` will give us all the shards of the first node, then all the shards of the second node, etc. The iteration order here is different: We take a shard from each node in succession. I've added docs describing the iteration order.

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 9 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: The hash code is rarely used on the data nodes (which is why I prefer to calculate it lazily). Now that ShardRouting is immutable we can use the racy single-check idiom (which is used in the JDK to calculate `java.lang.String#hashCode()`) without using volatile reads.

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 10 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: good catch. I removed the methods and made the stream constructors public.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 12 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: right!

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java (hunk 13 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: removed it.

**Comment by ywelsch** [comment]
@s1monw I've pushed a change addressing your review comments. Please have another look.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java (hunk 7 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: ah I remember that - nevermind good to add a comment it's not obvious

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 9 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: good lets do that

**Comment by s1monw** [comment]
LGTM thanks for all the javadocs

**Comment by ywelsch** [comment]
Thanks for the review @s1monw!
